File: pipes/submissions/accepted/pipes.py
from collections import deque

n = int(raw_input())
F = [int(i) for i in raw_input().split()]

group = [0] * n
children = [[] for _ in range(n)]

rootcount = 0
for i in range(n):
    fi = F[i]
    if (fi != -1):
        children[fi].append(i)
    else:
        rootcount += 1
        group[i] = rootcount

bestdrops = [None] * (rootcount + 1)
visited = [False] * n

for i in range(n):

    if visited[i] or group[i] == 0:
        continue

    bfs = deque([])
    bfs.append((0, i))
    visited[i] = True
    bestdrop = (n, n)

    while len(bfs) > 0:

        d, u = bfs.popleft()

        if len(children[u]) == 0:
            bestdrop = min(bestdrop, ( d, u ))

        for c in children[u]:
            if not visited[c]:
                group[c] = group[i]
                visited[c] = True
                bfs.append(( d + 1, c ))

    bestdrops[group[i]] = bestdrop[1]

q = int(raw_input())
S = [int(i) for i in raw_input().split()]
for i in range(q):
    s = S[i]
    print(bestdrops[group[s]])

File: pipes/submissions/accepted/Pipes.java
import java.util.*;


public class Pipes {
    static class Drop {
        int d;
        int u;
        public Drop(int d, int u) { this.d = d; this.u = u; }
        public Drop min(Drop other) {
            if (d < other.d) return this;
            else if (d == other.d) return u < other.u ? this : other;
            else return other;
        }
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        int n = scan.nextInt();
        int[] F = new int[n];
        int[] group = new int[n];
        LinkedList<Integer> children[] = new LinkedList[n];

        for (int i = 0; i < n; i++) {
            F[i] = scan.nextInt();
            children[i] = new LinkedList<>();
        }

        int rootcount = 0;
        for (int i = 0; i < n; i++) {
            int fi = F[i];
            if (fi != -1)
                children[fi].add(i);
            else
                group[i] = ++rootcount;
        }

        int[] bestdrops = new int[rootcount + 1];
        boolean[] visited = new boolean[n];

        for (int i = 0; i < n; i++) {
            if (visited[i] || group[i] == 0) 
                continue;

            LinkedList<Drop> bfs = new LinkedList<Drop>();
            bfs.add(new Drop(0, i));
            visited[i] = true;
            Drop bestdrop = new Drop(n, n);
            
            while (bfs.size() > 0) {
                Drop curr = bfs.removeFirst();
                if (children[curr.u].size() == 0) {
                    bestdrop =  bestdrop.min(curr);
                }

                for (int c : children[curr.u]) {
                    if (!visited[c]) {
                        group[c] = group[i];
                        visited[c] = true;
                        bfs.add(new Drop(curr.d + 1, c));
                    }
                }
            }

            bestdrops[group[i]] = bestdrop.u;
        }

        StringBuilder sb = new StringBuilder();
        int q = scan.nextInt();
        for (int i = 0; i < q; i++) {
            int s = scan.nextInt();
            sb.append(bestdrops[group[s]] + "\n");
        }
        System.out.print(sb.toString());
    }
}
